### DISCUSSION TITLE
How do you add SwiftLint into an Xcode project?

### DISCUSSION THREAD

**Discussion by cordechasse** [MobileNativeFoundation/rules_xcodeproj#2963] [Q&A] [answered]
Hi,

How can add SwiftLint into a xcodeproj?

In a "regular" XCode project, I would add a Run Script Phase with 
```
if which swiftlint > /dev/null; then
  swiftlint
else
  echo "warning: SwiftLint not installed, download from https://github.com/realm/SwiftLint"
fi
```
I'm not able to do the same with a xcodeproj bazel rule.

In my BUILD.bazel file I have :
```
xcodeproj(
    name = "xcodeproj",
    build_mode = "bazel",
    project_name = "iOSApp",
    tags = ["manual"],
    top_level_targets = [
        ":App",
    ],
    pre_build = "$SRCROOT/App/swiftlint.sh",
    extra_files = ["swiftlint"],
)

filegroup(
    name = "swiftlint",
    srcs = ["swiftlint.sh"],
    visibility = ["//visibility:public"],
)
```

It doesn't work, I get the following error : 

/private/var/tmp/_bazel_nicolasbush/63bf14c1f1d6f26e2f612fc11c487faa/rules_xcodeproj.noindex/build_output_base/external/rules_swift~/tools/worker/BUILD:113:10: Compiling tools/worker/worker_main.cc [for tool] failed: absolute path inclusion(s) found in rule '@@rules_swift~//tools/worker:worker':

So here my questions : 

1. The script should be in the "Run Script Phase" and I can only add to to the pre-build Run Script.
```
 pre_build = "$SRCROOT/App/swiftlint.sh",
 extra_files = ["swiftlint"],
```

Is there a way to add a Run Script Phase? I can't find any corresponding parameter in the xcodeproj method.

2. Is there another way? I found examples on Internet but it seems it's more to develop new swiftlint rules than to add SwiftLint to the XCode project.

Thanks for your answers.

**Comment by jpsim** [comment, marked as answer]
This is a question about SwiftLint, but these instructions are applicable for running any other script or CLI tool on pre-build Xcode actions.

We use SwiftLint as part of a pre-build phase for our `rules_xcodeproj` project and set it up this way:

First, we have a shell script wrapper that can download and run a prebuilt SwiftLint binary pinned to a specific version, so that devs don't all have to install it via Homebrew where the versions can easily get out of sync.

It looks like this:

```bash
#!/bin/bash

set -euo pipefail

readonly swiftlint_version="0.53.0"
readonly swiftlint_version_sha="8a949e9e8113ba0f60d69275e444a3ec78adb9598f2b93579bcaf2bb568950e5"
readonly swiftlint_version_url="https://github.com/realm/swiftlint/releases/download/$swiftlint_version/portable_swiftlint.zip"
script_root="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)"
readonly swiftlint_root="$script_root/tmp/swiftlint/versions"
readonly swiftlint="$swiftlint_root/swiftlint-$swiftlint_version"

if [[ ! -x "$swiftlint" ]]; then
  echo "Installing SwiftLint..." >&2
  mkdir -p $swiftlint_root

  download_swiftlint() {
    readonly swiftlint_zip="$swiftlint_root/portable_swiftlint.zip"
    curl --fail -L --retry 5 --retry-connrefused --silent --progress-bar \
      --output "$swiftlint_zip" "$swiftlint_version_url"
    unzip -p "$swiftlint_zip" swiftlint >"$swiftlint"
    rm -f "$swiftlint.zip"
  }

  download_swiftlint || download_swiftlint
  if echo "$swiftlint_version_sha  $swiftlint" | shasum --check --status; then
    chmod +x "$swiftlint"
  else
    echo "swiftlint sha mismatch" >&2
    shasum -a 256 "$swiftlint"
    rm -f "$swiftlint"
    exit 1
  fi
fi

exec "$swiftlint" "$@"
```

So we can run it as `./scripts/swiftlint version` or with whatever other arguments you want.

Then in our `rules_xcodeproj` configuration for our project, we have this:

```python
xcodeproj(
    name = "xcodeproj",
    pre_build = """set -euo pipefail

if [[ "$ACTION" == "build" ]]; then
  cd "$SRCROOT"
  ./scripts/swiftlint --quiet --lenient
fi""",
    project_name = "Ramp",
    tags = ["manual"],
    # etc...
)
```

This works great for us, and SwiftLint violations render correctly and quickly in the Xcode source editor and issue navigator.

<img width="789" alt="image" src="https://github.com/MobileNativeFoundation/rules_xcodeproj/assets/474794/1ba4c712-5e83-430d-9635-3a80dd6a7246">

This is just one way to do it. You could also run SwiftLint through Bazel using [skylib's `native_binary()` rule](https://github.com/bazelbuild/bazel-skylib/blob/main/docs/native_binary_doc.md#native_binary) or any other number of ways.

**Comment by BalestraPatrick** [comment]
The above answer by @jpsim is pretty similar to what we do. One difference given our project is very big is that we use the git information to only lint the changed files, but that's totally optional and based on your needs. So you could modify your `pre_build` with this:
```
# Get all files which have been modified. If nothing has been modified, exit right away. This will ignore untracked files.
all_modified_files+=$(git diff --name-only --staged --diff-filter=d && git ls-files --others --exclude-standard --full-name --modified)
[ -z "$all_modified_files" ] && exit 0

# Filter by staged files, and only get swift files needed for swiftlint
swift_files+=$(echo "$all_modified_files" | grep \.swift$ | tr '\n' ' ')

# Check if we have any modified swift files, if we do we can use swiftlint
if [ -n "$swift_files" ]; then
   # For local development, convert errors as warnings for speed of iteration.
   swiftlint lint --config .swiftlint.yml --quiet $swift_files | sed "s/error:/warning:/"
fi
```

**Comment by cordechasse** [comment]
Thanks a lot for your replies, it's very helpful.

It's that kind of advice that helps new users to start using Bazel. You rocks!
